### DISCUSSION TITLE
Uses CIDR database within relations

### DISCUSSION THREAD

**Discussion by ikus060** [sqlalchemy/sqlalchemy#7989] [Usage Questions] [answered]
Hello,

I want to create a relationship between two Table. One are IP Addresses and the other are Subnets define using IP_CIDR.

```
class Subnet(Base):
    ip_cidr = Column(CidrType, unique=True, nullable=False)

class IP(Base):
    ip = Column(String, unique=True, nullable=False)
    related_subnets = relationship(
        Subnet,
        primaryjoin=lambda: Subnet.ip_cidr.supernet_of(Ip.ip),
        viewonly=True,
        lazy=True,
    )
```

`CidrType` is a custom datatype I've define to wrap postgresql CIDR.

Source could be found here: https://gitlab.com/ikus-soft/udb/-/blob/master/src/udb/core/model/_cidr.py

When defining the relation as bellow, I'm always getting errors:
```
Failed with Error: [undefined]failed on setup with "sqlalchemy.exc.ArgumentError: Could not locate any relevant foreign key columns for primary join condition 'supernet_of(subnet.ip_cidr, anon_1.ip)' on relationship Ip.related_subnets.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation."
cls = <class 'test_network.IpTest'>
```

I'm guessing the issue is related to the custom datatype and the call use `supernet_of` but without a clue how to get this fixed. If even supported by SQLAlchemy.

My Goal is to get a SQL similar to the following:

```
SELECT *
FROM subnet
WHERE subnet.ip_cidr >> ?
```

**Comment by zzzeek** [comment, marked as answer]
there's a documented example of more or less the same thing using CIDR at https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#using-custom-operators-in-join-conditions
